llvm-external-projects/iree-dialects/lib/Dialect/LinalgExt/IR/LinalgExtOps.cpp

```diff
@@ -1447,6 +1484,388 @@ LogicalResult TopkOp::getResultTilePosition(
+    return op->emitError(
+        "blocking factors must equal the number of dimensions to block");
+
+  // The number of `dim_pos` that carries the index of the dimensions to block
+  // must be less or equal than the input rank.
+  if (dimsPos.size() > getInputRank())
+    return op->emitError(
+        "indices of blocked dimensions must be less or equal than "
+        "the input rank");
+  if (isInvalid(dimsPos))
+    return op->emitError("invalid dims_pos vector");
+  // Require `dim_pos` to be in-bound. `dim_pos` carries the index of the
+  // dimensions to block.
+  if (!isInBound(dimsPos, getOutputRank()))
+    return op->emitError("out-of-bound position");
+
+  // Require output rank to match input rank + number of blocking factors.
+  if ((getInputRank() + numberOfBlockingFactors) != getOutputRank())
+    return op->emitError(
+        "output rank must equal input rank + blocking factors");
+
+  // Verify tiles. Make sure each provided tile is non-zero.
+  if (hasZeros(getMixedTiles()))
+    return op->emitError("invalid tile factor");
+
+  // Bail out if the tile does not divide the dimension fully. In the case of
+  // dynamic tile factors or dimensions, having a partial tile is undefined
+  // behavior. We will relax this constraint when we introduce padding
+  // semantics.
+  if (areNotFullTiles(getInputShape(), getDimAndStaticTileMapping()))
+    return op->emitError(
+        "invalid tile factor provided. Only full tiles are supported");
+
+  // Verify result type against inferred type.
+  ShapedType expectedType = PackOp::inferResultType();
+  if (expectedType != getOutputType())
+    return op->emitError(
+               "inferred type do not match provied output type. Expected ")
+           << expectedType << " but got: " << getOutputType();
+
+  return success();
+}
+
+// Get the tile sizes as `OpFoldResult`.
+SmallVector<OpFoldResult> PackOp::getMixedTiles() {
+  SmallVector<OpFoldResult> mixedInnerTiles;
+  mixedInnerTiles.reserve(getInputRank());
+  unsigned dynamicValIndex = 0;
+  for (Attribute attr : getStaticInnerTiles()) {
+    auto tileAttr = attr.cast<IntegerAttr>();
+    if (!ShapedType::isDynamic(tileAttr.getInt()))
+      mixedInnerTiles.push_back(tileAttr);
+    else
+      mixedInnerTiles.push_back(getInnerTiles()[dynamicValIndex++]);
+  }
+  return mixedInnerTiles;
+}
+
+// Return the tile sizes as `int64_t`. If a tile size is dynamic a sentinel
+// `kDynamicSize` is introduced at that position in the returned vector.
+SmallVector<int64_t> PackOp::getStaticTiles() {
+  SmallVector<Value> dynamicTiles;
+  SmallVector<int64_t> staticTiles;
+  dispatchIndexOpFoldResults(getMixedTiles(), dynamicTiles, staticTiles,
+                             ShapedType::kDynamicSize);
+  return staticTiles;
+}
+
+// Implement the tiling interface. The number of loops equals
+// the rank of the output tensors. All the loops are parallel.
+SmallVector<StringRef> PackOp::getLoopIteratorTypes() {
```

Comment:
I have to think about how this works out in practice. If we want this op to be "tilable", I expect the number of iterations to be same as the dimensionality of the input tensor rather than the output tensor.